castai/resource_gke_cluster_id.go

```diff
@@ -0,0 +1,188 @@
+		},
+
+		Schema: map[string]*schema.Schema{
+			FieldGKEClusterIdName: {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+				Description:      "GKE cluster name",
+			},
+			FieldGKEClusterIdProjectId: {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+				Description:      "GCP project id",
+			},
+			FieldGKEClusterIdLocation: {
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotWhiteSpace),
+				Description:      "GCP cluster zone in case of zonal or region in case of regional cluster",
+			},
+			FieldClusterToken: {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+				Description: "CAST.AI agent cluster token",
+			},
+			FieldGKEClientSA: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Service account email in client project",
+			},
+			FieldGKECastSA: {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: "Service account email in cast project",
+			},
+		},
+	}
+}
+
+func resourceCastaiGKEClusterIdCreate(ctx context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ProviderConfig).api
+
+	req := sdk.ExternalClusterAPIRegisterClusterJSONRequestBody{
+		Name: data.Get(FieldGKEClusterName).(string),
+	}
+
+	location := data.Get(FieldGKEClusterLocation).(string)
+	region := location
+	// Check if location is zone or location.
+	if strings.Count(location, "-") > 1 {
+		// region "europe-central2"
+		// zone "europe-central2-a"
+		regionParts := strings.Split(location, "-")
+		regionParts = regionParts[:2]
+		region = strings.Join(regionParts, "-")
+	}
+
+	req.Gke = &sdk.ExternalclusterV1GKEClusterParams{
+		ProjectId:   toPtr(data.Get(FieldGKEClusterProjectId).(string)),
+		Region:      &region,
+		Location:    &location,
+		ClusterName: toPtr(data.Get(FieldGKEClusterName).(string)),
+	}
+
+	log.Printf("[INFO] Registering new external cluster: %#v", req)
+	resp, err := client.ExternalClusterAPIRegisterClusterWithResponse(ctx, req)
+	if checkErr := sdk.CheckOKResponse(resp, err); checkErr != nil {
+		return diag.FromErr(checkErr)
+	}
+
+	clusterID := *resp.JSON200.Id
+	tkn, err := createClusterToken(ctx, client, clusterID)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if err := data.Set(FieldClusterToken, tkn); err != nil {
+		return diag.FromErr(fmt.Errorf("setting cluster token: %w", err))
+	}
+	data.SetId(clusterID)
+	// If client service account is set, create service account on cast side.
+	if len(data.Get(FieldGKEClientSA).(string)) > 0 {
```

Comment:
Shouldn't this be a separate resource?

Reply:
it is actually a new resource https://github.com/castai/terraform-provider-castai/pull/386/files#diff-a5a684b5b5f8e9cbaac3f6a93c0e15e642c09317befcb3e7e11c0a6fbb557f6aR33 I use it to create registered cluster first and initialize it with cast service account and service key.